File: backend/app/admin/schemas/mock_instance.py
from typing import Optional

from pydantic import Field

from ._base import CamelModel


class MockInstanceCreate(CamelModel):
    name: str = Field(..., min_length=1, max_length=100)
    topology_id: str = Field(..., min_length=1)
    port: int = Field(..., ge=1, le=65535)
    description: Optional[str] = Field(default=None, max_length=200)
    enabled: bool = True


class MockInstanceUpdate(CamelModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    topology_id: Optional[str] = None
    port: Optional[int] = Field(default=None, ge=1, le=65535)
    description: Optional[str] = Field(default=None, max_length=200)
    enabled: Optional[bool] = None


class MockInstanceItem(CamelModel):
    id: str
    name: str
    topology_id: str
    topology_name: str
    port: int
    description: Optional[str]
    enabled: bool
    api_count: int = 0
    created_at: str
    updated_at: str


class RequestLogItem(CamelModel):
    id: int
    ts: str
    api_id: Optional[str] = None
    method: str
    path: str
    query: Optional[str] = None
    status_code: int
    duration_ms: int
    client_ip: Optional[str] = None
    error_message: Optional[str] = None
    instance_id: Optional[str] = None

File: frontend/src/api/mockInstance.ts
import { apiGet, apiPost, apiPut, apiPatch, apiDelete } from './http'

export interface MockInstanceItem {
  id: string
  name: string
  topologyId: string
  topologyName: string
  port: number
  description: string | null
  enabled: boolean
  apiCount: number
  createdAt: string
  updatedAt: string
}

export interface MockInstanceCreate {
  name: string
  topologyId: string
  port: number
  description?: string | null
  enabled?: boolean
}

export interface MockInstanceUpdate {
  name?: string | null
  topologyId?: string | null
  port?: number
  description?: string | null
  enabled?: boolean
}

export const mockInstanceApi = {
  list: (): Promise<{ items: MockInstanceItem[]; total: number }> =>
    apiGet('/mock-instances'),

  create: (data: MockInstanceCreate): Promise<{ id: string }> =>
    apiPost('/mock-instances', data),

  update: (id: string, data: MockInstanceUpdate): Promise<{ id: string }> =>
    apiPut(`/mock-instances/${id}`, data),

  delete: (id: string): Promise<{ id: string }> =>
    apiDelete(`/mock-instances/${id}`),

  patchEnabled: (id: string, enabled: boolean): Promise<{ id: string }> =>
    apiPatch(`/mock-instances/${id}/enabled`, { enabled }),
}

export interface RequestLogItem {
  id: number
  ts: string
  apiId: string | null
  method: string
  path: string
  query: string | null
  statusCode: number
  durationMs: number
  clientIp: string | null
  errorMessage: string | null
  instanceId: string | null
}

export interface RequestLogResponse {
  items: RequestLogItem[]
  hasMore: boolean
}

export const requestLogApi = {
  fetchLogs: (instId: string, params?: { limit?: number; before?: string }): Promise<RequestLogResponse> =>
    apiGet(`/mock-instances/${instId}/logs`, params),

  clearLogs: (instId: string): Promise<{ id: string }> =>
    apiDelete(`/mock-instances/${instId}/logs`),
}
